source/extensions/http/cache/file_system_http_cache/file_system_http_cache.cc

```diff
@@ -388,6 +405,78 @@ void FileSystemHttpCache::trackFileRemoved(uint64_t file_size) {
   stats_.size_bytes_.set(size_bytes_);
 }
 
+void FileSystemHttpCache::maybeEvict() {
+  uint64_t size_to_evict = 0;
+  uint64_t count_to_evict = 0;
+  if (config().has_max_cache_size_bytes()) {
+    // capture a value from the atomic because we're going to use it twice and don't want our
+    // value to change in between.
+    uint64_t seen_size = size_bytes_;
+    size_to_evict = seen_size > config().max_cache_size_bytes().value()
+                        ? seen_size - config().max_cache_size_bytes().value()
+                        : 0;
+  }
+  if (config().has_max_cache_entry_count()) {
+    // capture a value from the atomic because we're going to use it twice and don't want our
+    // value to change in between.
+    uint64_t seen_count = size_count_;
+    count_to_evict = seen_count > config().max_cache_entry_count().value()
+                         ? seen_count - config().max_cache_entry_count().value()
+                         : 0;
+  }
+  if (size_to_evict == 0 && count_to_evict == 0) {
+    return;
+  }
+
+  auto os_sys_calls = Api::OsSysCallsSingleton::get();
+  uint64_t size = 0;
+  uint64_t count = 0;
+  uint64_t proposed_size_evicted = 0;
+  struct ProposedEviction {
+    std::string name_;
+    uint64_t size_;
+    struct timespec last_touch_;
+    bool operator<(const ProposedEviction& other) const { return last_touch_ < other.last_touch_; }
+  };
+  std::multiset<ProposedEviction> proposed_evictions;
+
+  for (const Filesystem::DirectoryEntry& entry : Filesystem::Directory(std::string{cachePath()})) {
+    if (!isCacheFile(entry)) {
+      continue;
+    }
+    count++;
+    size += entry.size_bytes_.value_or(0);
+    struct stat s;
+    if (os_sys_calls.stat(absl::StrCat(cachePath(), entry.name_).c_str(), &s).return_value_ != -1) {
+      struct timespec last_touch = std::max(s.st_atim, s.st_ctim);
```

Comment:
i've never used it but it might be worth considering the C++17 support for this: https://en.cppreference.com/w/cpp/filesystem/last_write_time

Reply:
We *might* get "last read or written" from stat, which is preferable behavior over just last written. There doesn't seem to be a C++17 function for that.